File: Leetcode_Medium/_55. Jump Game/Solution.py
class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        # The farthest index that can be reached
        max_reachable = 0
        
        # Iterate through each index in the array
        for i in range(len(nums)):
            # If the current index is beyond the maximum reachable index, return False
            if i > max_reachable:
                return False
            # Update the maximum reachable index if the current index plus its value is greater
            max_reachable = max(max_reachable, i + nums[i])
            # If the maximum reachable index is beyond or at the last index, return True
            if max_reachable >= len(nums) - 1:
                return True
        
        # If the loop finishes without reaching the last index, return False
        return False

File: Leetcode_Medium/_55. Jump Game/Solution.java
public class Solution {
    public boolean canJump(int[] nums) {
        // The farthest index that can be reached
        int maxReachable = 0;
        
        // Iterate through each index in the array
        for (int i = 0; i < nums.length; i++) {
            // If the current index is beyond the maximum reachable index, return false
            if (i > maxReachable) {
                return false;
            }
            // Update the maximum reachable index if the current index plus its value is greater
            maxReachable = Math.max(maxReachable, i + nums[i]);
            // If the maximum reachable index is beyond or at the last index, return true
            if (maxReachable >= nums.length - 1) {
                return true;
            }
        }
        
        // If the loop finishes without reaching the end of the array, return false
        return false;
    }
}

public class Solution {
    public boolean canJump(int[] nums) {
        int maxReachable = 0;
        int lastIndex = nums.length - 1; // Store the last index
        
        // Iterate through each index up to maxReachable
        for (int i = 0; i <= maxReachable && maxReachable < lastIndex; i++) {
            maxReachable = Math.max(maxReachable, i + nums[i]);
        }
        
        // Return true if maxReachable reaches or exceeds the last index
        return maxReachable >= lastIndex;
    }
}